Port DH-11692: Add native support for java.time types

nbauernfeind opened the pull request:
I did a bit more replication than was in the DHE implementation; but hopefully for the better.

DH-11692 via commits:
- https://github.com/deephaven-ent/iris/commit/563e69e16a2206b0c90ec220b0bd75f4a68e3bac
- https://github.com/deephaven-ent/iris/commit/f680b26788f1d4ab998e374278eff81449c85b7f
- https://github.com/deephaven-ent/iris/commit/4f46299190e396093860e2a9becd1382f21ffd39
- https://github.com/deephaven-ent/iris/commit/7baba9f419c6ab9f2fb0b16c6ad77da414b28237

Follow up work:
- #3367
- #3359

Nightlies: https://github.com/nbauernfeind/deephaven-core/actions/runs/4058490234

cpwright commented on engine/time/src/main/java/io/deephaven/time/DateTimeUtils.java:
```diff
@@ -1240,128 +1234,182 @@ public static Clock currentClock() {
+     * @param timeZone The {@link TimeZone} to use for the conversion.
+     * @return A {@link ZonedDateTime} using the specified time zone. or null if dateTime was null
      */
-    public static String lastBusinessDateNy() {
```
Removing lastBusinessDateNy() is probably defensible if you do not want to own calendars at all (but then Enterprise needs to reimplement something).  You'll also lose a lot of plot utility w/o business calendars.

I think not having currentDateNy() which does not rely on calendars is pretty bad.  @rbasralian do you think this is a tenable position with respect to usability?

None of the override stuff I particular see us needing.

nbauernfeind replied:
As discussed offline, for the foreseeable future, DHE will own supporting clients in this regard.

nbauernfeind commented on engine/table/src/main/java/io/deephaven/engine/table/impl/util/TableTimeConversions.java:
```diff
@@ -0,0 +1,341 @@
+ *
+ * <pre>
+ * baseTable = db.i("Market", "Trades")
+ *               .where("Date &#62; 2021-10-01")
```
Hmm. In the pre-block this seems to render just fine using `<`.

rcaudy commented on engine/table/src/main/java/io/deephaven/engine/table/impl/sources/regioned/RegionedColumnSourceInstant.java:
```diff
@@ -0,0 +1,96 @@
+/**
+ * Regioned column source implementation for columns of {@link Instant}s.
+ */
+final class RegionedColumnSourceInstant
```
Can you make this class and its peers a little more mutually consistent? (`elementIndex`->`rowKey`, `final @NotNull` -> `@NotNull final`, spacing, etc)

nbauernfeind commented on Util/src/main/java/io/deephaven/util/type/TypeUtils.java:
```diff
@@ -512,14 +513,15 @@ public static boolean isCharacter(@NotNull final Class<?> c) {
     public static boolean isDateTime(Class<?> type) {
-        return Date.class.isAssignableFrom(type) || type.getAnnotation(IsDateTime.class) != null
-                && type.getAnnotation(IsDateTime.class).value();
+        return Date.class.isAssignableFrom(type)
```
This method is used to create time filters and to determine if the column is displayable to the user. (If it is non-displayable, then it gets `toString`d.) It is only used by server-side UI helpers.

rcaudy replied:
Why not ZonedDateTime, or LocalXYZ?

rcaudy replied:
We discussed that `Date` probably does not belong here, other code that thinks `Serializable` works should be changed, and you need to debug `Instant` and or `ZonedDateTime`.

nbauernfeind commented on engine/table/src/main/java/io/deephaven/engine/table/impl/by/LongChunkedAddOnlyMinMaxOperator.java:
```diff
@@ -42,7 +49,15 @@ class LongChunkedAddOnlyMinMaxOperator implements IterativeChunkedAggregationOpe
         this.name = name;
         // region resultColumn initialization
-        resultColumn = type == DateTime.class ? new DateTimeArraySource() : new LongArraySource();
+        if (type == DateTime.class) {
```
It probably would be nice. As noted offline we're probably in trouble anywhere we lose the ZoneId (say via `ComponentType` usages). I've opened an issue for creating a `ColumnType` that we can pass around instead of `Class<?>`. This is probably a lot better than introducing an optional `ZoneId` to pass through multiple layers.

See #3455.

nbauernfeind commented on Util/src/main/java/io/deephaven/util/codec/ZonedDateTimeCodec.java:
```diff
@@ -0,0 +1,96 @@
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+
+public class ZonedDateTimeCodec implements ObjectCodec<ZonedDateTime> {
```
Hmm, it's hard for me to tell. I'm not sure how to utilize it, but it does seem like there could exist parquet metadata that would utilize this codec if referenced by class name.

nbauernfeind commented on Util/src/main/java/io/deephaven/util/codec/ZonedDateTimeCodec.java:
```diff
@@ -0,0 +1,96 @@
+        return safeComputeNanos(value.toEpochSecond(), value.getNano());
+    }
+
+    private static long safeComputeNanos(long epochSecond, long nanoOfSecond) {
```
Note that this exception is preventing `epochSecond` from overflowing as well. I've added `Math.addExact` for the summation, though.

rcaudy commented on engine/table/src/main/java/io/deephaven/engine/table/impl/by/AggregationProcessor.java:
```diff
@@ -507,6 +510,9 @@ final void addMinOrMaxOperator(final boolean isMin, @NotNull final String inputN
                     return;
                 }
             }
+            if (rawInputSource instanceof ConvertableTimeSource.Zoned) {
+                ZoneId id = ((ConvertableTimeSource.Zoned) rawInputSource).getZone();
+            }
```
Seems like a change that didn't pan out?

rcaudy commented on engine/table/src/test/java/io/deephaven/engine/table/impl/sources/AbstractBooleanColumnSourceTest.java:
```diff
@@ -21,8 +21,11 @@
 import io.deephaven.engine.rowset.RowSetBuilderSequential;
 import io.deephaven.engine.rowset.RowSetFactory;
 import io.deephaven.engine.rowset.chunkattributes.RowKeys;
+import io.deephaven.engine.table.ChunkSink;
```
All of these new imports are unused. That's OK, if these are a replication artifact.

nbauernfeind commented:
I can't seem to reply to your final comment. I added ZonedDateTime/Instant to the list. I removed Date. For now LocalDate and LocalTime are stubbed with FixedSizeBinary formats, but these are not implemented between server/client neither for worker-to-worker nor for the jsapi. 

At this moment, barrage strips the ZoneId from ZonedDateTime columns and sends as UTC. The ser/deser code acts on Class<?> and it would be much nicer to use ColumnDataType instead of passing extra args.

rcaudy commented on engine/table/src/main/java/io/deephaven/engine/table/impl/sources/ReinterpretUtils.java:
```diff
@@ -98,6 +99,68 @@ public static ColumnSource<Long> zonedDateTimeToLongSource(final @NotNull Column
+    }
+
+    /**
+     * Given an {@link Instant} column source turn it into a long column source, either via reinterpretation or
+     * wrapping.
```
Remove `or wrapping` from this one and `writableInstantToLongSource`.
Maybe consider mentioning that the source is writable on all 5 new methods? Feel not strictly necessary, since the types are specified.

rcaudy commented on engine/table/src/main/java/io/deephaven/engine/table/impl/sources/ReinterpretUtils.java:
```diff
@@ -98,6 +99,68 @@ public static ColumnSource<Long> zonedDateTimeToLongSource(final @NotNull Column
+     * @return the long source or null if it could not be reinterpretted
+     */
+    @NotNull
+    public static WritableColumnSource<Long> writableZonedDateTimeToLongSource(
+            final @NotNull WritableColumnSource<?> source) {
+        if (source.allowsReinterpret(long.class)) {
+            return (WritableColumnSource<Long>) source.reinterpret(long.class);
+        }
+        return null;
+    }
```
You have 3 identical method bodies. Converge them?

rcaudy commented on engine/table/src/main/java/io/deephaven/engine/table/impl/sources/InMemoryColumnSource.java:
```diff
@@ -18,6 +18,7 @@
 import org.jetbrains.annotations.Nullable;
 
 import java.time.Instant;
+import java.time.ZonedDateTime;
```
Unused

rcaudy commented on engine/table/src/main/java/io/deephaven/engine/table/impl/select/analyzers/SelectColumnLayer.java:
```diff
@@ -110,7 +110,13 @@ final public class SelectColumnLayer extends SelectOrViewColumnLayer {
+            if (dataSource.getType() != writableSource.getType()) {
+                // this should only occur when using primitives internally and the user has requested a non-primitive
+                chunkSource = ReinterpretUtils.maybeConvertToPrimitive(dataSource);
+            } else {
```
We should follow the reinterpret with an `Assert` that the resulting types are now equal.
